packages/mdc-snackbar/util.js

```diff
@@ -0,0 +1,88 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+import {numbers, strings} from './constants';
+
+const {ARIA_LIVE_DELAY_MS} = numbers;
+const {ARIA_LIVE_LABEL_TEXT_ATTR} = strings;
+
+/**
+ * @param {!HTMLElement} ariaEl
+ * @param {!HTMLElement=} labelEl
+ */
+export function announce(ariaEl, labelEl = ariaEl) {
+  const priority = ariaEl.getAttribute('aria-live');
+  const labelText = labelEl.textContent.trim(); // Ignore `&nbsp;` (see below)
+  if (!labelText) {
+    return;
+  }
+
+  // Temporarily disable `aria-live` to prevent JAWS+Firefox from announcing the message twice.
+  ariaEl.setAttribute('aria-live', 'off');
+
+  // Temporarily clear `textContent` to force a DOM mutation event that will be detected by screen readers.
+  // `aria-live` elements are only announced when the element's `textContent` *changes*, so snackbars
+  // sent to the browser in the initial HTML response won't be read unless we clear the element's `textContent` first.
+  // Similarly, displaying the same snackbar message twice in a row doesn't trigger a DOM mutation event,
+  // so screen readers won't announce the second message unless we first clear `textContent`.
+  //
+  // We have to clear the label text two different ways to make it work in all browsers and screen readers:
+  //
+  //   1. `textContent = ''` is required for IE11 + JAWS
+  //   2. `innerHTML = '&nbsp;'` is required for Chrome + JAWS and NVDA
+  //
+  // All other browser/screen reader combinations support both methods.
+  //
+  // The wrapper `<span>` visually hides the space character so that it doesn't cause jank when added/removed.
+  // N.B.: Setting `position: absolute`, `opacity: 0`, or `height: 0` prevents Chrome from detecting the DOM change.
+  //
+  // This technique has been tested in:
+  //
+  //   * JAWS 18.0 & 2019:
+  //       - Chrome 70
+  //       - Firefox 60 (ESR)
+  //       - IE 11
+  //   * NVDA 2017 & 2018:
+  //       - Chrome 70
+  //       - Firefox 60 (ESR)
+  //       - IE 11
+  labelEl.textContent = '';
+  labelEl.innerHTML = '<span style="display: inline-block; width: 0; height: 1px;">&nbsp;</span>';
+
+  // Prevent visual jank by temporarily displaying the label text in the ::before pseudo-element.
+  // CSS generated content is normally announced by screen readers
+  // (except in IE 11; see https://tink.uk/accessibility-support-for-css-generated-content/);
+  // however, `aria-live` is turned off, so this DOM update will be ignored by screen readers.
+  labelEl.setAttribute(ARIA_LIVE_LABEL_TEXT_ATTR, labelText);
+
+  setTimeout(() => {
```

Comment:
I'm a little wary that we're not storing / returning the handle on this timeout, but maybe I'm being paranoid. Is there a chance of the snackbar being dismissed before this timeout even elapses and causing weird behavior? Or will it simply not announce because the snackbar will become display: none?

Reply:
If it's `display: none` when the timeout fires, the browsers I tested (Chrome/IE11/Firefox in NVDA/JAWS) don't announce it.

The timeout is responsible for resetting the `textContent` back to its original value, so I'm not sure we'd want to clear it. But there could be a use case I'm not thinking of.

I made sure to test for non-empty `textContent` at the top of the function, so if you call it twice in a row, the second call is a noop.

From my testing, Windows screen readers queue the message and read it fully as long as the `textContent` was reset to its original value _before_ the snackbar was set to `display: none`.